File: apps/mobile/test/contracts/fabrary_printing_match_contract_test.dart
import 'package:fabtrades/core/logic/fabrary_csv.dart';
import 'package:fabtrades/core/logic/fabrary_match.dart';
import 'package:flutter_test/flutter_test.dart';

import 'contract_fixtures.dart';

void main() {
  final contract = loadContract('fabrary_printing_match');
  final cases = contractCases(contract, 'cases');

  group('fabrary printing match contract', () {
    for (final testCase in cases) {
      test(testCase['name'] as String, () {
        var row = Map<String, dynamic>.from(testCase['row'] as Map);
        final csv = testCase['csv'] as String?;
        if (csv != null) {
          final parsed = parseFabraryCsv(csv);
          expect(parsed.ok, isTrue);
          expect(parsed.rows.first['Name'], row['Name']);
          row = Map<String, dynamic>.from(parsed.rows.first);
        }
        final catalog = (testCase['catalog'] as List)
            .map((e) => Map<String, dynamic>.from(e as Map))
            .toList();
        final result = matchFabraryRow(row, catalog);
        expect(result.toJson(), testCase['expected']);
      });
    }
  });
}

File: apps/web/tests/contracts/fabraryPrintingMatch.contract.test.js
import contract from '../../../../packages/contracts/fabrary_printing_match.json';
import { parseFabraryCsv } from '../../src/utils/fabraryCsv.js';
import { matchFabraryRow } from '../../src/utils/fabraryMatch.js';

describe('fabrary printing match contract', () => {
    contract.cases.forEach((testCase) => {
        it(testCase.name, () => {
            let row = testCase.row;
            if (testCase.csv) {
                const parsed = parseFabraryCsv(testCase.csv);
                expect(parsed.ok).toBe(true);
                expect(parsed.rows[0].Name).toBe(testCase.row.Name);
                row = parsed.rows[0];
            }
            expect(matchFabraryRow(row, testCase.catalog)).toEqual(testCase.expected);
        });
    });
});
